ports/qemu-arm/Makefile

```diff
@@ -86,6 +147,11 @@ repl: $(BUILD)/firmware.elf
 run: $(BUILD)/firmware.elf
 	$(QEMU_SYSTEM) $(QEMU_ARGS) -serial pty -kernel $<
 
+# `make debug` will block QEMU until a debugger is connected to port 1234.
+.PHONY: debug
+debug: $(BUILD)/firmware.elf
+	$(QEMU_SYSTEM) $(QEMU_ARGS) -serial mon:stdio -S -s -kernel $<
+
 .PHONY: test
 test: $(BUILD)/firmware.elf
```

Comment:
I'd add a `testdebug` target too.  How would you be able to see things like #15589 in a debugger otherwise?

Reply:
It's not possible to used both `-S` and the `run-tests.py` infrastructure, because the latter consumes stdio of qemu-system-X and so you won't be able to start the emulator because the monitor is not accessible.

> That issue was reproduced by running several tests in a row -

I think we could make this `debug` target use `-serial pty` and then you can run the tests manually as many times as you like against the exposed serial device, eg:
```
$ make debug
<in a separate terminal>
$ gdb ...
<in a separate terminal>
$ ./run-tests.py --target qemu-arm --device /dev/pts/123
```

Reply:
I just tried the version I committed a few days ago as https://github.com/agatti/micropython/commit/ce8a877310dba2c09fcf985758a524ed3f7cbb79, using `-serial pty`, and it seems to work when running `make testdebug` in the first terminal and when it's ready spawning another one and run `gdb` attaching to the qemu instance.  Unless something else changed there's no need to use three terminals :)

Reply:
> But what I'm also saying above is that it's possible to rerun the test suite over and over, or whatever combination of tests you like. That requires 3 terminals.

Oh, never figured that was a possibility.  Almost every time I've had the need for this it turned out to be a fatal crash so there was no point in rerunning the test without restarting qemu as well.  In that case then yes, you're absolutely right.

Reply:
You are right, you can use `-S -s` with the `make test` target.

Instead of duplicating all targets with a debug version, I added an orthogonal option `QEMU_DEBUG` which can be enabled for any qemu target, eg:
```
$ make QEMU_DEBUG=1 test
```
That will run the test suite with `-S -s`.

I also added `QEMU_DEBUG_EXTRA`, and updated the README with descriptions of these options.